File: src/server/functions/discovery.ts
import { createServerFn } from '@tanstack/react-start'
import { discoverAPIs, groupResources } from '../kube/discovery'

export const getAPIResources = createServerFn({ method: 'GET' }).handler(
  async () => {
    const resources = await discoverAPIs()
    return {
      resources,
      groups: groupResources(resources),
    }
  },
)

File: backend/app/routers/discovery.py
from fastapi import APIRouter

from app.kube.discovery import discover_apis, group_resources
from app.models import DiscoveryResponse

router = APIRouter(prefix="/api/resources", tags=["discovery"])


@router.get("/discover", response_model=DiscoveryResponse)
async def get_api_resources() -> DiscoveryResponse:
    resources = await discover_apis()
    return DiscoveryResponse(resources=resources, groups=group_resources(resources))
